Wferrer feature #155877650 reset password

jerazo opened the pull request:
- Add a hasRole test on entity instead, not on role repo to test if a user has a given role
- Added check on preUpdate and created preCreate to hinder creation of reset password entry if no role and email not validated
- Wrote unit test

victoriousgaming committed:
Adding last entity as an easy way for the events to look at the last entity made

victoriousgaming committed:
Ready to test

jerazo committed:
- password reset tests
- notification still failing

victoriousgaming committed:
Trying saving raven in a different spot

victoriousgaming committed:
Fixing small problems like broken paths

victoriousgaming committed:
Merge branch 'develop' of https://github.com/tempestwf/tempest-tools-common into develop

# Conflicts:
#	src/Laravel/Http/Middleware/BasicDataExtractorMiddleware.php

victoriousgaming committed:
Fixing more broken paths

victoriousgaming committed:
Meaningless comit about the path for raven package

victoriousgaming committed:
Updating group name for tests

victoriousgaming committed:
Updating group name for tests

victoriousgaming committed:
Small clean up

victoriousgaming committed:
Progress but no solution to fix the issue

victoriousgaming committed:
Tests run all the way through

victoriousgaming committed:
Tests run all the way through with raven on

victoriousgaming committed:
Messages no work, but leave behind a messy DB

victoriousgaming committed:
Making it give an error if you try to set a verified token to false

jerazo committed:
- Add a hasRole test on entity instead, not on role repo to test if a user has a given role
- Added check on preUpdate and created preCreate to hinder creation of reset password entry if no role and email not validated
- Wrote unit test

tempestwf commented on app/API/V1/Entities/User.php:
```diff
@@ -229,6 +236,11 @@ public function getRoles():?Collection
 
+    public function hasRole()
+    {
+        return !!$this->getRoles();
```
Shouldn't this check for a specific role? It needs to see if the user has the 'user' role specifically

jerazo replied:
Changed

tempestwf commented on app/API/V1/Repositories/PasswordResetRepository.php:
```diff
@@ -0,0 +1,136 @@
+
+        /* check if has role */
+        if (!$user->hasRole()) {
+            throw PasswordResetException::noRole();
```
This should check for the specific role

jerazo replied:
Changed

tempestwf commented on app/API/V1/Repositories/PasswordResetRepository.php:
```diff
@@ -0,0 +1,136 @@
+        $params = $k['params'];
+        /** @var PasswordReset $entity */
+        $roleRepo = $this->getEm()->getRepository(User::class);
+        $user = $roleRepo->find(array_keys($params['user']['read'])[0]);
```
Note that this also will get refactored due to some changes I put on the ticket. Generally, it would be better to get the entity out of $k['entity']->getUser() (provided that is populated at this point).

jerazo replied:
preCreate does not have $k['entity'] though.

tempestwf replied:
@jerazo in that case I think your solution will work, except I had another task in the ticket for not passing in the userId at all, but just passing in an email as an option, and then we find the user from the email. This is more secure.

tempestwf commented on app/API/V1/Repositories/PasswordResetRepository.php:
```diff
@@ -0,0 +1,136 @@
+        }
+
+        /* check if has role */
+        if (!$entity->getUser()->hasRole()) {
```
Should check for the specific role

jerazo replied:
Changed

tempestwf reviewed (changes requested):
Found some changes to make. Also I left some refactors in the ticket that I realized we need for added security.

jerazo committed:
- hasRole to check specific role
- hasRole to be uses on addUserRoles
- PR cleanup

jerazo committed:
- Password reset stable
- Unit test
- .env updated

jerazo committed:
- phpunit.xml

jerazo committed:
- #157212480: All tokens should expire
- password reset and email verification default to 1 day expiration (set in .env per min)

jerazo committed:
- cleanup other tests with respect to email verification and password reset changes.

tempestwf commented on app/API/V1/Entities/PasswordReset.php:
```diff
@@ -0,0 +1,202 @@
+     * @ORM\PreUpdate
+     * @throws \RuntimeException
+     */
+    public function ttPrePersist():void
```
Since we have successfully phased this out we can remove it.

tempestwf reviewed (changes requested):
I also left one question in slack. Other wise looks great, nice work on what turned out to be a fairly large ticket.
